### DISCUSSION TITLE
Two issues about Drainage and InvasionPercolation in version 3.10

### DISCUSSION THREAD

**Discussion by Aneurinsky** [PMEAL/OpenPNM#2659] [Q&A]
Hi, The openpnm open source package has enhanced my understanding of two-phase flow at the pore scale. In the study of the new version, there are **two questions** that bother me, and I still need your answers:

（1）
 In version 3.10, the **Drainage** algorithm is selected and the calculation is completed, and the file is saved using **project_to_vtk**, and the program reports an error:
![1](https://user-images.githubusercontent.com/121919936/210487569-846f00c7-0312-4d48-a5c3-d0cc1109868c.png)

Similarly, when the **InvasionPercolation** algorithm is selected, an error will also appear when saving the file with **project_to_vtk**:
![2](https://user-images.githubusercontent.com/121919936/210487638-6900f66e-fda2-424d-aa1f-94c569752934.png)

（2）
When applying the **Drainage** algorithm, the **intrusion phase** is usually indicated, so is there a **default defense phase**? If so, what is this **defense phase**? If not, was the initial state of the network dry or vacuum?

Similarly, for the **InvasionPercolation** algorithm, if the **defense phase** is not specified, what is the **default defense phase**? In **RelativePermeability.py**, set the **defense phase** to the default value, or to air, the calculation result is the same? I understand the following statement, the air is not set as a **defense phase**, I don't know if this understanding is correct?

`air = op.phase.Air(network=pn,name='air')
air['pore.surface_tension'] = 0.072
air['pore.contact_angle'] = 180.0
air.add_model_collection(op.models.collections.phase.air)
air.add_model_collection(op.models.collections.physics.basic)
air.regenerate_models()

water = op.phase.Water(network=pn,name='water')
water.add_model_collection(op.models.collections.phase.water)
water.add_model_collection(op.models.collections.physics.basic)
water.regenerate_models()

DRN = op.algorithms.Drainage(network=pn,phase=water)
DRN.set_inlet_BC(pores=pn.pores('top'))
DRN.run()`

Thank you again for your patience in reading my question and look forward to your reply, thank you！

**Comment by jgostick** [comment]
Hi @Aneurinsky 
The first issue you mention seems like a bug.  The file io function is looking for an attribute on the algorithm that is not there.  I will investigate and open an issue asap.

To answer your second question(s) about the defending phase, during quasi-static invasions (slow), the defending phase properties do not impact the displacement since there is no viscous effects so there is no need to specific a defending phase.  However, trapping can still occur so if you specify 'outlets' then the algorithm will check to see which defending phase pores are no longer connected to the outlets and mark them as trapped.  If you don't specify outlets then the algorithm assumes that trapping does not occur, which is equivalent to assuming the defending phase is a vacuum, like in mercury porosimetry.

**Reply by Aneurinsky** [reply]
Thank you for the prompt response, professor. Your team is so nice...

**Reply by JahanzebTariq** [reply]
@Aneurinsky... would you please also paste the code through which you are trying to save the percolation pathways. thanks in advance

**Reply by JahanzebTariq** [reply]
@jgostick .. I'm also having the same issue

**Reply by jgostick** [reply]
Just pushed a fix.  Will merge into dev today.  There is one other branch which I would like to also merge into the dev before creating a new release.  This other branch is also almost ready, so I expect to have a new "patch" release this weeks, like OpenPNM 3.1.2.

**Reply by JahanzebTariq** [reply]
I am still getting this issue please tell me if this has been resolved or not..? thank you
![image](https://user-images.githubusercontent.com/99289270/214763110-50491ec3-8e9e-405f-9f66-0e3301911651.png)

**Reply by jgostick** [reply]
The MacOS builds are failing, so I was waiting to figure that out before merge.  The patch is available on the `dev` branch if you are able to use that.  It requires pulling the source code from Github, installing using `pip install -e /path/to/cloned/repo`, then you 'checkout' whichever branch you like.

**Reply by JahanzebTariq** [reply]
professor i'm not that good at these things.. approximately how long is it going to take for the bug to be completely fixed at your end?

**Comment by JahanzebTariq** [comment]
Greeting @jgostick .. SIr I've extracted the percolation pathways and want to visualize them in paraview..
![image](https://user-images.githubusercontent.com/99289270/217328018-1ce19502-6a3f-448a-a1a5-65c406dd429e.png)
but this issue has not been resolved yet..
![image](https://user-images.githubusercontent.com/99289270/217328327-a347bc90-5897-473e-8a23-5133f24c62f2.png)

please address this issue as soon as possible. I would be very grateful for this.. thank you in advance

**Reply by jgostick** [reply]
Try this as a work-around:

``ip.soln = {'pore.invasion_pressure': None}``

This will put the missing attribute on your algorithm and hopefully bypass the check. 

> Just a reminder that this was fixed on the dev branch within days of you reporting it, so the problem is your unwillingness to learn how to use git to access the latest code. Since you're doing modelling-based research I strongly suggest you get over this hurdle...it's not that difficult to pull some code from github and run it, and it will look good on your resume if you know how to do it.

**Reply by JahanzebTariq** [reply]
@jgostick.. sorry for not being able to understand you.. I'll try the dev solution that you told me. Thank you for knocking some sense into me..

**Reply by JahanzebTariq** [reply]
@jgostick... would you please tell me in a little more detail how to do it.. My sincere apologies.. but if you guide me a little more I'll be able to get through nicely.. please do reply

**Reply by jgostick** [reply]
I'm sorry, I should have checked my suggestion first.  I have now confirmed that the following will work:

``` python
drn.settings['quantity'] = 'throat.invasion_pressure'
drn.soln = {'throat.invasion_pressure': drn['throat.invasion_pressure']}
```

**Reply by JahanzebTariq** [reply]
@jgostick ... sir should I go with this one or should I do the dev version?

**Reply by JahanzebTariq** [reply]
![image](https://user-images.githubusercontent.com/99289270/217556243-5f1a57a3-df22-4da7-aa0a-25f766b5a60c.png)
 @jgostick .. sir did you tell me the solution to this issue? please do confirm

**Reply by JahanzebTariq** [reply]
@jgostick .. did what you said but still the error remains
![image](https://user-images.githubusercontent.com/99289270/217561336-6e1ed679-c72f-4171-b7f1-afe559ebedb6.png)

sorry to be such a pain but please do guide me if i am wrong.. thank you
